internal/pkg/dl/artifact.go

```diff
@@ -0,0 +1,65 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package dl
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+
+	"github.com/elastic/fleet-server/v7/internal/pkg/bulk"
+	"github.com/elastic/fleet-server/v7/internal/pkg/dsl"
+	"github.com/elastic/fleet-server/v7/internal/pkg/model"
+)
+
+const (
+	artifactsIndexName = ".fleet-artifacts"
+)
+
+var (
+	QueryArtifactBySha2 = prepareQueryArtifactBySha2()
+	ErrConflict         = errors.New("Fail multiple artifacts for the same sha256")
+)
+
+func prepareQueryArtifactBySha2() *dsl.Tmpl {
+	root := dsl.NewRoot()
+	tmpl := dsl.NewTmpl()
+
+	root.Query().Bool().Filter().Term(FieldEncodedSha256, tmpl.Bind(FieldEncodedSha256), nil)
+	tmpl.MustResolve(root)
+	return tmpl
+}
+
+func FindArtifactBySha256(ctx context.Context, bulker bulk.Bulk, sha2 string) (*model.Artifact, error) {
+
+	res, err := SearchWithOneParam(
+		ctx,
+		bulker,
+		QueryArtifactBySha2,
+		artifactsIndexName,
+		FieldEncodedSha256,
+		sha2,
+	)
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(res.Hits) == 0 {
+		return nil, ErrNotFound
+	}
+
+	if len(res.Hits) > 1 {
```

Comment:
If this happens, I assume it means there is a bug in the Kibana code?

Reply:
Yes.  The sha2 is a primary key.  We really do not want more than one sha2 record in the index; it becomes unclear which is the correct record to pull.
But as I think about it, it could in theory happen due to multiple Kibana's trying to create the same artifact in the same way at the same time.  This is due to not using the sha2 to derive the documentId.  In addition, could be the case that N artifacts have the same body but are of different types, causing a collision.

Will work with @paul-tavares  on how to best address.

Reply:
Ok, after talking with Paul, will make a couple of changes:

1) Change the Fleet-Server code do a term query based on identity+sha2, instead of just sha2 to avoid the case where there are two records with the same payload, but different identities.
2) Tolerate record dupes in the Fleet Server.  We will log when we see more than one, pick the first and continue.
3) Validate the sha2 in the Fleet Server.  This is just good hygiene.
4) Consider using the sha2+identity to derive the document id in Kibana.  This will avoid race conditions across Kibanas as well as dupes.

Reply:
SGTM. I assume Kibana takes care of cleaning up "old" artifacts. We should make sure when artifacts are deleted that also all the dupes are cleaned up.

Reply:
Yep.  Spoke with @paul-tavares about this.  It is a bit complicated given that multiple policies will point to the same artifact records.  The plan is to implement a garbage collector that runs every N hours, interprets the current policies, and removes any artifacts that are no longer referenced.